scrapy/downloadermiddlewares/httpauth.py

```diff
@@ -15,17 +16,28 @@ class HttpAuthMiddleware:
 
     @classmethod
     def from_crawler(cls, crawler):
-        o = cls()
+        o = cls()   
         crawler.signals.connect(o.spider_opened, signal=signals.spider_opened)
         return o
 
     def spider_opened(self, spider):
-        usr = getattr(spider, 'http_user', '')
-        pwd = getattr(spider, 'http_pass', '')
-        if usr or pwd:
-            self.auth = basic_auth_header(usr, pwd)
+        pass
+
+    def _authenticated(self, response):
+        if response.status == "401":
+            return False
+
+        return True
 
     def process_request(self, request, spider):
-        auth = getattr(self, 'auth', None)
-        if auth and b'Authorization' not in request.headers:
+        if _authenticated(self, Response(request.url)):
+            if request.url in spider.url_user_pwd.keys():
```

Comment:
Should this be checking the domain instead? With this code, slightly different URLs that point to the same resource will need duplicated entries. Also, the `keys()` call is not necessary.

Reply:
Not sure but I can check if its working with domains.